lib/synapse/service_watcher/zookeeper_poll.rb

```diff
@@ -0,0 +1,77 @@
+require 'synapse/service_watcher/base'
+require 'synapse/service_watcher/zookeeper'
+
+require 'zk'
+require 'thread'
+
+class Synapse::ServiceWatcher
+  class ZookeeperPollWatcher < ZookeeperWatcher
+    def start
+      log.info 'synapse: ZookeeperPollWatcher starting'
+
+      @should_exit = false
+      @thread = nil
+      @poll_interval = @discovery['polling_interval_sec']
+
+      zk_connect do
+        @thread = Thread.new {
+          log.info 'synapse: zookeeper polling thread started'
+
+          # last_run is shifted by a random jitter in order to spread the first
+          # discover call of multiple Synapses. This helps to spread load.
+          # As long as the beginning is spread, the future discovers will also
+          # be spread.
+          last_run = Time.now - rand(@poll_interval)
+
+          until @should_exit
+            now = Time.now
+            elapsed = now - last_run
+
+            if elapsed >= @poll_interval
+              last_run = now
+              discover
+            end
+
+            sleep 0.5
```

Comment:
why sleep 0.5?

Reply:
The idea here is to have an infinite loop in a background thread that can terminate quickly. We check whether or not to exit every iteration, so `sleep 0.5` caps the termination time to ~0.5 seconds.

If instead I do `sleep @poll_interval`, the thread could remain alive for a very long time potentially, depending on whatever the setting is.

Reply:
make sense